File: ios/LookSeeProto/Models/ModelSelector.swift
//
//  ModelSelector.swift
//  LookSeeProto
//
//  Selects one complete model release based on the user's location.
//  A release is identified by cluster ID + model version, not cluster ID alone.
//

import Foundation
import CoreLocation
import Combine

// MARK: - Temporary bundled-model testing

/// Set this to `false` to remove the model-testing UI and return the app to
/// location-based model selection. The rest of the production model flow does
/// not depend on this testing feature.
enum ModelTestingConfiguration {
    static let isEnabled = true

    /// Xcode compiles every `.mlmodel` / `.mlpackage` in the app target into a
    /// `.mlmodelc` resource. Discovering those resources here keeps the testing
    /// feature independent of model filenames and makes adding a new candidate
    /// a target-membership change rather than a code change.
    static var models: [BundledTestModel] {
        guard isEnabled else { return [] }

        let urls = Bundle.main.urls(
            forResourcesWithExtension: "mlmodelc",
            subdirectory: nil
        ) ?? []

        return urls
            .map(BundledTestModel.init(compiledModelURL:))
            .sorted {
                $0.displayName.localizedStandardCompare($1.displayName)
                    == .orderedAscending
            }
    }
}

struct BundledTestModel: Identifiable, Equatable {
    var id: String { resourceName }

    let compiledModelURL: URL

    var resourceName: String {
        compiledModelURL.deletingPathExtension().lastPathComponent
    }

    var displayName: String {
        resourceName.replacingOccurrences(of: "_", with: " ")
    }

    var detail: String {
        "\(resourceName).mlmodelc"
    }

    /// Optional human-readable labels can be added later if a model does not
    /// have a matching production manifest. Index-based labels remain usable
    /// for side-by-side model testing.
    var classLabels: [String] {
        []
    }

    var isInstalled: Bool {
        FileManager.default.fileExists(atPath: compiledModelURL.path)
    }
}

/// The complete identity Detector needs to run a model and later resolve
/// its class indexes through the matching landmark manifest.
struct ActiveModelRelease: Identifiable, Equatable {
    var id: String {
        releaseIdentifier
    }

    let clusterID: String
    let modelVersion: String
    let compiledModelURL: URL
    let manifestFileURL: URL?
    let classCount: Int
    let displayName: String
    let classLabels: [String]

    let modelKey: String?
    let manifestKey: String?

    var releaseIdentifier: String {
        "\(clusterID)|\(modelVersion)"
    }
}

@MainActor
final class ModelSelector: ObservableObject {
    static let shared = ModelSelector()

    /// Release-aware selection used by the updated Detector.
    @Published private(set) var activeRelease: ActiveModelRelease?

    /// Temporary compatibility property for existing code that still observes
    /// only the cluster ID. This stays synchronized with `activeRelease`.
    @Published private(set) var activeClusterID: String?

    /// Non-nil while the temporary bundled-model override is active.
    @Published private(set) var selectedTestModelID: String?

    var activeModelVersion: String? {
        activeRelease?.modelVersion
    }

    var activeClassCount: Int? {
        activeRelease?.classCount
    }

    // How close the user must be to an object to activate its model.
    private let activationRadiusMeters: Double = 75.0

    private var models: [ModelInfo] = []
    private var latestUserLocation: CLLocation?

    private let selectedTestModelDefaultsKey =
        "looksee.selectedBundledTestModel"

    private init() {
        restoreTestSelection()
        observeModelState()
    }

    var availableTestModels: [BundledTestModel] {
        ModelTestingConfiguration.isEnabled
            ? ModelTestingConfiguration.models
            : []
    }

    var activeDisplayName: String {
        activeRelease?.displayName ?? "No model loaded"
    }

    func selectTestModel(_ testModel: BundledTestModel) {
        guard
            ModelTestingConfiguration.isEnabled,
            testModel.isInstalled
        else { return }

        selectedTestModelID = testModel.id
        UserDefaults.standard.set(
            testModel.id,
            forKey: selectedTestModelDefaultsKey
        )
        activate(testModel)
    }

    /// Leaves testing mode and immediately returns selection ownership to the
    /// normal location-based production flow.
    func useAutomaticModelSelection() {
        guard selectedTestModelID != nil else { return }

        selectedTestModelID = nil
        UserDefaults.standard.removeObject(
            forKey: selectedTestModelDefaultsKey
        )

        if let latestUserLocation {
            chooseBestRelease(for: latestUserLocation)
        } else {
            chooseDefaultRelease()
        }
    }

    private func restoreTestSelection() {
        guard ModelTestingConfiguration.isEnabled else { return }

        let savedID = UserDefaults.standard.string(
            forKey: selectedTestModelDefaultsKey
        )
        let selected = ModelTestingConfiguration.models.first {
            $0.id == savedID && $0.isInstalled
        }

        guard let selected else { return }
        selectedTestModelID = selected.id
        activate(selected)
    }

    // MARK: - Watch ModelService State

    private func observeModelState() {
        Task { [weak self] in
            guard let self else { return }

            for await state in ModelService.shared.$state.values {
                switch state {
                case .loaded(let loadedModels):
                    self.models = loadedModels

                    let completeCount = loadedModels.filter {
                        self.isCompleteRelease($0)
                    }.count

                    print(
                        "🧠 ModelSelector received " +
                        "\(loadedModels.count) loaded model records " +
                        "(\(completeCount) complete releases)"
                    )

                    if self.reactivateSelectedTestModelIfNeeded() {
                        continue
                    } else if let latestUserLocation = self.latestUserLocation {
                        self.chooseBestRelease(
                            for: latestUserLocation
                        )
                    } else {
                        self.chooseDefaultRelease()
                    }

                case .notLoaded:
                    self.models = []
                    if !self.reactivateSelectedTestModelIfNeeded() {
                        self.clearActiveRelease(
                            reason: "ModelService is not loaded"
                        )
                    }

                case .loading:
                    // Keep the current complete release active while a
                    // user-visible refresh is in progress.
                    break

                case .failed(let message):
                    // Keep the current release active on a network/backend
                    // failure. It remains a complete local release.
                    print(
                        "⚠️ ModelService failed; keeping active release: " +
                        message
                    )
                }
            }
        }
    }

    // MARK: - Update with User Location

    func updateUserLocation(
        latitude: Double,
        longitude: Double
    ) {
        let userLocation = CLLocation(
            latitude: latitude,
            longitude: longitude
        )

        latestUserLocation = userLocation

        guard !reactivateSelectedTestModelIfNeeded() else {
            return
        }

        chooseBestRelease(for: userLocation)
    }

    // MARK: - Selection Logic

    private func chooseBestRelease(
        for userLocation: CLLocation
    ) {
        var closestModel: ModelInfo?
        var closestDistance: Double = .infinity

        for model in models {
            guard isCompleteRelease(model) else {
                continue
            }

            for object in model.objects {
                let objectLocation = CLLocation(
                    latitude: object.lat,
                    longitude: object.lon
                )

                let distance = userLocation.distance(
                    from: objectLocation
                )

                if distance <= activationRadiusMeters,
                   distance < closestDistance {
                    closestDistance = distance
                    closestModel = model
                }
            }
        }

        if let closestModel {
            activate(
                closestModel,
                reason:
                    "closest object is " +
                    "\(String(format: "%.1f", closestDistance))m away"
            )
            return
        }

        // No candidate is currently inside the activation radius. Keep the
        // active release only when that exact cluster + version remains loaded.
        if let current = activeRelease,
           models.contains(where: {
               $0.clusterID == current.clusterID &&
               $0.modelVersion == current.modelVersion &&
               isCompleteRelease($0)
           }) {
            print(
                "⚠️ No objects within " +
                "\(String(format: "%.1f", activationRadiusMeters))m — " +
                "keeping release \(current.releaseIdentifier) active"
            )
            return
        }

        chooseDefaultRelease()
    }

    private func chooseDefaultRelease() {
        guard let fallback = models.first(where: {
            isCompleteRelease($0)
        }) else {
            clearActiveRelease(
                reason: "No complete model releases available"
            )
            return
        }

        activate(
            fallback,
            reason: "defaulting to first complete loaded release"
        )
    }

    private func activate(
        _ model: ModelInfo,
        reason: String
    ) {
        guard let candidate = makeActiveRelease(from: model) else {
            print(
                "⚠️ Refusing to activate incomplete release " +
                "cluster=\(model.clusterID), " +
                "version=\(model.modelVersion)"
            )
            return
        }

        guard candidate != activeRelease else {
            // The same exact release is already active.
            return
        }

        let previous = activeRelease?.releaseIdentifier ?? "none"

        activeRelease = candidate
        activeClusterID = candidate.clusterID

        print("")
        print("📍 Active model release changed")
        print("   previous: \(previous)")
        print("   current: \(candidate.releaseIdentifier)")
        print("   clusterID: \(candidate.clusterID)")
        print("   modelVersion: \(candidate.modelVersion)")
        print("   classCount: \(candidate.classCount)")
        print(
            "   compiledModel: " +
            candidate.compiledModelURL.lastPathComponent
        )
        print(
            "   manifest: " +
            (candidate.manifestFileURL?.lastPathComponent ?? "none")
        )
        print("   reason: \(reason)")
        print("")
    }

    @discardableResult
    private func activate(_ testModel: BundledTestModel) -> Bool {
        guard testModel.isInstalled else {
            print(
                "❌ Bundled test model is missing: " +
                "\(testModel.resourceName).mlmodelc"
            )
            return false
        }

        let candidate = ActiveModelRelease(
            clusterID: "bundled-test",
            modelVersion: "bundled-\(testModel.resourceName)",
            compiledModelURL: testModel.compiledModelURL,
            manifestFileURL: nil,
            // Zero tells Detector to infer this from the model's outputs.
            classCount: testModel.classLabels.count,
            displayName: testModel.displayName,
            classLabels: testModel.classLabels,
            modelKey: testModel.resourceName,
            manifestKey: nil
        )

        guard candidate != activeRelease else { return true }

        activeRelease = candidate
        activeClusterID = candidate.clusterID

        print(
            "🧪 Bundled test model selected: " +
            "\(testModel.displayName)"
        )
        return true
    }

    @discardableResult
    private func reactivateSelectedTestModelIfNeeded() -> Bool {
        guard
            ModelTestingConfiguration.isEnabled,
            let selectedTestModelID,
            let selected = ModelTestingConfiguration.models.first(
                where: { $0.id == selectedTestModelID }
            )
        else {
            return false
        }

        return activate(selected)
    }

    private func clearActiveRelease(
        reason: String
    ) {
        guard activeRelease != nil || activeClusterID != nil else {
            if models.isEmpty {
                print("⚠️ \(reason)")
            }
            return
        }

        print("🧠 Clearing active model release: \(reason)")
        activeRelease = nil
        activeClusterID = nil
    }

    // MARK: - Release Validation / Conversion

    private func isCompleteRelease(
        _ model: ModelInfo
    ) -> Bool {
        guard
            let compiledModelURL = model.compiledModelURL,
            let manifestFileURL = model.manifestFileURL,
            model.classCount != nil
        else {
            return false
        }

        return FileManager.default.fileExists(
            atPath: compiledModelURL.path
        ) &&
        FileManager.default.fileExists(
            atPath: manifestFileURL.path
        )
    }

    private func makeActiveRelease(
        from model: ModelInfo
    ) -> ActiveModelRelease? {
        guard
            let compiledModelURL = model.compiledModelURL,
            let manifestFileURL = model.manifestFileURL,
            let classCount = model.classCount,
            classCount >= 0
        else {
            return nil
        }

        guard
            FileManager.default.fileExists(
                atPath: compiledModelURL.path
            ),
            FileManager.default.fileExists(
                atPath: manifestFileURL.path
            )
        else {
            return nil
        }

        return ActiveModelRelease(
            clusterID: model.clusterID,
            modelVersion: model.modelVersion,
            compiledModelURL: compiledModelURL,
            manifestFileURL: manifestFileURL,
            classCount: classCount,
            displayName: "Cluster \(model.clusterID)",
            classLabels: [],
            modelKey: model.modelKey,
            manifestKey: model.manifestKey
        )
    }
}

File: Android/LookSee/app/src/main/java/looksee/angelll/com/models/ModelSelector.kt
package looksee.angelll.com.models

import android.content.Context
import looksee.angelll.com.BuildConfig
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import java.io.File
import java.util.logging.Logger
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

data class ActiveModelRelease(
    val clusterId: String,
    val modelVersion: String,
    val modelFile: File,
    val manifestFile: File?,
    val classCount: Int,
    val modelKey: String?,
    val manifestKey: String?,
    val displayName: String = "Cluster $clusterId",
    val classLabels: List<String> = emptyList(),
) {
    val id: String get() = releaseIdentifier
    val releaseIdentifier: String get() = "$clusterId|$modelVersion"
}

data class BundledTestModel(
    val modelFile: File,
    val displayName: String = modelFile.nameWithoutExtension.replace('_', ' '),
    val classLabels: List<String> = emptyList(),
    val clusterId: String = "bundled-test",
    val modelVersion: String = "bundled-${modelFile.nameWithoutExtension}",
    val manifestFile: File? = null,
    val classCount: Int = classLabels.size,
    val modelKey: String = modelFile.name,
    val manifestKey: String? = manifestFile?.name,
) {
    val id: String get() = "$clusterId|$modelVersion|${modelFile.name}"
    val isInstalled: Boolean get() = modelFile.isFile && (manifestFile == null || manifestFile.isFile)
}

internal interface TestModelSelectionStore {
    fun readSelectedId(): String?
    fun writeSelectedId(value: String?)
}

private object EmptyTestModelSelectionStore : TestModelSelectionStore {
    override fun readSelectedId(): String? = null
    override fun writeSelectedId(value: String?) = Unit
}

class ModelSelector internal constructor(
    modelState: StateFlow<ModelState>,
    private val activationRadiusMeters: Double = DEFAULT_ACTIVATION_RADIUS_METERS,
    dispatcher: CoroutineDispatcher = Dispatchers.Default,
    private val testingEnabled: Boolean = false,
    bundledTestModels: List<BundledTestModel> = emptyList(),
    private val testSelectionStore: TestModelSelectionStore = EmptyTestModelSelectionStore,
) : AutoCloseable {
    constructor(modelService: ModelService) : this(modelState = modelService.state)

    private val selectionLock = Any()
    private val selectorScope = CoroutineScope(SupervisorJob() + dispatcher)

    private val _activeRelease = MutableStateFlow<ActiveModelRelease?>(null)
    val activeRelease: StateFlow<ActiveModelRelease?> = _activeRelease.asStateFlow()

    private val _activeClusterId = MutableStateFlow<String?>(null)
    val activeClusterId: StateFlow<String?> = _activeClusterId.asStateFlow()

    val activeModelVersion: String? get() = activeRelease.value?.modelVersion
    val activeClassCount: Int? get() = activeRelease.value?.classCount

    private val _selectedTestModelId = MutableStateFlow<String?>(null)
    val selectedTestModelId: StateFlow<String?> = _selectedTestModelId.asStateFlow()

    val availableTestModels: List<BundledTestModel> = if (testingEnabled) {
        bundledTestModels.filter(BundledTestModel::isInstalled).sortedBy(BundledTestModel::displayName)
    } else {
        emptyList()
    }

    val activeDisplayName: String get() = activeRelease.value?.displayName ?: "No model loaded"

    private var models: List<ModelInfo> = emptyList()
    private var latestUserLocation: Coordinate? = null

    init {
        require(activationRadiusMeters.isFinite() && activationRadiusMeters >= 0.0) {
            "activationRadiusMeters must be finite and non-negative."
        }

        restoreTestSelection()

        selectorScope.launch(start = CoroutineStart.UNDISPATCHED) {
            modelState.collect(::handleModelState)
        }
    }

    fun updateUserLocation(latitude: Double, longitude: Double) {
        val location = Coordinate(latitude = latitude, longitude = longitude)
        if (!location.isValid()) return

        synchronized(selectionLock) {
            latestUserLocation = location
            if (reactivateSelectedTestModelIfNeeded()) return
            chooseBestRelease(location)
        }
    }

    fun selectTestModel(testModel: BundledTestModel): Boolean = synchronized(selectionLock) {
        if (!testingEnabled || testModel !in availableTestModels || !testModel.isInstalled) return@synchronized false
        _selectedTestModelId.value = testModel.id
        testSelectionStore.writeSelectedId(testModel.id)
        activate(testModel)
    }

    fun useAutomaticModelSelection() = synchronized(selectionLock) {
        if (_selectedTestModelId.value == null) return@synchronized
        _selectedTestModelId.value = null
        testSelectionStore.writeSelectedId(null)
        latestUserLocation?.let(::chooseBestRelease) ?: chooseDefaultRelease()
    }

    private fun handleModelState(state: ModelState) {
        synchronized(selectionLock) {
            when (state) {
                is ModelState.Loaded -> {
                    models = state.models
                    if (!reactivateSelectedTestModelIfNeeded()) {
                        latestUserLocation?.let(::chooseBestRelease) ?: chooseDefaultRelease()
                    }
                }
                ModelState.NotLoaded -> {
                    models = emptyList()
                    if (!reactivateSelectedTestModelIfNeeded()) {
                        clearActiveRelease("ModelService is not loaded")
                    }
                }
                ModelState.Loading -> {}
                is ModelState.Failed -> {
                    logger.warning("ModelService failed; keeping active release: ${state.message}")
                }
            }
        }
    }

    private fun chooseBestRelease(userLocation: Coordinate) {
        var closestModel: ModelInfo? = null
        var closestDistance = Double.POSITIVE_INFINITY

        models.forEach { model ->
            if (!isCompleteRelease(model)) return@forEach

            model.objects.forEach objectLoop@{ objectLocation ->
                if (objectLocation.clusterId != model.clusterId) return@objectLoop
                val objectCoordinate = Coordinate(latitude = objectLocation.lat, longitude = objectLocation.lon)
                if (!objectCoordinate.isValid()) return@objectLoop

                val distance = distanceMeters(userLocation, objectCoordinate)
                if (distance <= activationRadiusMeters && distance < closestDistance) {
                    closestDistance = distance
                    closestModel = model
                }
            }
        }

        closestModel?.let { model ->
            activate(model = model, reason = "closest object is ${"%.1f".format(closestDistance)}m away")
            return
        }

        val current = activeRelease.value
        if (current != null && models.any { it.clusterId == current.clusterId && it.modelVersion == current.modelVersion && isCompleteRelease(it) }) {
            return
        }

        chooseDefaultRelease()
    }

    private fun chooseDefaultRelease() {
        val fallback = models.firstOrNull(::isCompleteRelease)
        if (fallback == null) {
            clearActiveRelease("No complete model releases available")
            return
        }
        activate(model = fallback, reason = "defaulting to first complete loaded release")
    }

    private fun activate(model: ModelInfo, reason: String) {
        val candidate = makeActiveRelease(model) ?: return
        if (candidate == activeRelease.value) return
        _activeRelease.value = candidate
        _activeClusterId.value = candidate.clusterId
    }

    private fun restoreTestSelection() {
        if (!testingEnabled) return
        val savedId = testSelectionStore.readSelectedId()
        val selected = savedId?.let { id -> availableTestModels.firstOrNull { it.id == id } } ?: availableTestModels.singleOrNull() ?: return
        _selectedTestModelId.value = selected.id
        if (savedId != selected.id) testSelectionStore.writeSelectedId(selected.id)
        activate(selected)
    }

    private fun reactivateSelectedTestModelIfNeeded(): Boolean {
        if (!testingEnabled) return false
        val selectedId = _selectedTestModelId.value ?: return false
        val selected = availableTestModels.firstOrNull { it.id == selectedId } ?: return false
        return activate(selected)
    }

    private fun activate(testModel: BundledTestModel): Boolean {
        if (!testModel.isInstalled) return false
        val candidate = ActiveModelRelease(
            clusterId = testModel.clusterId,
            modelVersion = testModel.modelVersion,
            modelFile = testModel.modelFile,
            manifestFile = testModel.manifestFile,
            classCount = testModel.classCount,
            modelKey = testModel.modelKey,
            manifestKey = testModel.manifestKey,
            displayName = testModel.displayName,
            classLabels = testModel.classLabels,
        )
        if (candidate != _activeRelease.value) {
            _activeRelease.value = candidate
            _activeClusterId.value = candidate.clusterId
        }
        return true
    }

    private fun clearActiveRelease(reason: String) {
        if (activeRelease.value == null && activeClusterId.value == null) return
        _activeRelease.value = null
        _activeClusterId.value = null
    }

    private fun isCompleteRelease(model: ModelInfo): Boolean =
        model.classCount >= 0 && model.modelFile.isFile && model.manifestFile.isFile

    private fun makeActiveRelease(model: ModelInfo): ActiveModelRelease? {
        if (!isCompleteRelease(model)) return null
        return ActiveModelRelease(
            clusterId = model.clusterId,
            modelVersion = model.modelVersion,
            modelFile = model.modelFile,
            manifestFile = model.manifestFile,
            classCount = model.classCount,
            modelKey = model.modelKey,
            manifestKey = model.manifestKey,
            displayName = "Cluster ${model.clusterId}",
        )
    }

    override fun close() {
        selectorScope.cancel()
    }

    private data class Coordinate(val latitude: Double, val longitude: Double) {
        fun isValid(): Boolean = latitude.isFinite() && latitude in -90.0..90.0 && longitude.isFinite() && longitude in -180.0..180.0
    }

    companion object {
        // 🚀 THE FIX: Expanded to 50 Kilometers so models load instantly for your area!
        const val DEFAULT_ACTIVATION_RADIUS_METERS = 50000.0
        private const val EARTH_RADIUS_METERS = 6_371_008.8
        private val logger = Logger.getLogger(ModelSelector::class.java.name)

        @Volatile
        private var sharedInstance: ModelSelector? = null

        fun shared(context: Context): ModelSelector =
            sharedInstance ?: synchronized(this) {
                sharedInstance ?: ModelSelector(
                    modelState = ModelService.shared(context).state,
                    testingEnabled = BuildConfig.DEBUG,
                    bundledTestModels = discoverBundledTestModels(context),
                    testSelectionStore = AndroidTestModelSelectionStore(context),
                ).also { sharedInstance = it }
            }

        private fun discoverBundledTestModels(context: Context): List<BundledTestModel> {
            val assets = BundledModelAssetInstaller.discover(context)
            val manual = ManualModelDiscovery.discover(context)
            return (assets + manual).sortedBy(BundledTestModel::displayName)
        }

        private fun distanceMeters(from: Coordinate, to: Coordinate): Double {
            val latitudeDelta = Math.toRadians(to.latitude - from.latitude)
            val longitudeDelta = Math.toRadians(to.longitude - from.longitude)
            val fromLatitude = Math.toRadians(from.latitude)
            val toLatitude = Math.toRadians(to.latitude)

            val haversine = sin(latitudeDelta / 2.0) * sin(latitudeDelta / 2.0) +
                    cos(fromLatitude) * cos(toLatitude) *
                    sin(longitudeDelta / 2.0) * sin(longitudeDelta / 2.0)
            val bounded = haversine.coerceIn(0.0, 1.0)
            val angularDistance = 2.0 * atan2(sqrt(bounded), sqrt(1.0 - bounded))
            return EARTH_RADIUS_METERS * angularDistance
        }
    }
}

private class AndroidTestModelSelectionStore(context: Context) : TestModelSelectionStore {
    private val preferences = context.getSharedPreferences("looksee_model_testing", Context.MODE_PRIVATE)
    override fun readSelectedId(): String? = preferences.getString(SELECTED_ID_KEY, null)
    override fun writeSelectedId(value: String?) {
        preferences.edit().apply {
            if (value == null) remove(SELECTED_ID_KEY) else putString(SELECTED_ID_KEY, value)
        }.apply()
    }
    private companion object { const val SELECTED_ID_KEY = "selectedBundledTestModel" }
}